core/hosting/HostWithCoreClrHost/src/SampleHost.cpp

```diff
@@ -0,0 +1,324 @@
+
+    // Construct the CoreCLR path
+    // For this sample, we know CoreCLR's path. For other hosts,
+    // it may be necessary to probe for coreclr.dll/libcoreclr.so
+    std::string coreClrPath(runtimePath);
+    coreClrPath.append(FS_SEPERATOR);
+    coreClrPath.append(CORECLR_FILE_NAME);
+
+    // Construct the managed library path
+    std::string managedLibraryPath(runtimePath);
+    managedLibraryPath.append(FS_SEPERATOR);
+    managedLibraryPath.append(MANAGED_ASSEMBLY);
+    
+    //
+    // STEP 1: Load CoreCLR (coreclr.dll/libcoreclr.so)
+    //
+#if WINDOWS
+    // <Snippet1>
+    HMODULE coreClr = LoadLibraryExA(coreClrPath.c_str(), NULL, 0);
+    // </Snippet1>
+#elif LINUX 
+    void *coreClr = dlopen(coreClrPath.c_str(), RTLD_NOW | RTLD_LOCAL);
+#endif 
+    if (coreClr == NULL)
+    {
+        printf("ERROR: Failed to load CoreCLR from %s\n", coreClrPath.c_str());
+        return -1;
+    }
+    else
+    {
+        printf("Loaded CoreCLR from %s\n", coreClrPath.c_str());
+    }
+
+    //
+    // STEP 2: Get CoreCLR hosting functions
+    //
+#if WINDOWS
+    // <Snippet2>
+    coreclr_initialize_ptr initializeCoreClr = (coreclr_initialize_ptr)GetProcAddress(coreClr, "coreclr_initialize");
+    coreclr_create_delegate_ptr createManagedDelegate = (coreclr_create_delegate_ptr)GetProcAddress(coreClr, "coreclr_create_delegate");
+    coreclr_shutdown_ptr shutdownCoreClr = (coreclr_shutdown_ptr)GetProcAddress(coreClr, "coreclr_shutdown");
+    // </Snippet2>
+#elif LINUX    
+    coreclr_initialize_ptr initializeCoreClr = (coreclr_initialize_ptr)dlsym(coreClr, "coreclr_initialize");
+    coreclr_create_delegate_ptr createManagedDelegate = (coreclr_create_delegate_ptr)dlsym(coreClr, "coreclr_create_delegate");
+    coreclr_shutdown_ptr shutdownCoreClr = (coreclr_shutdown_ptr)dlsym(coreClr, "coreclr_shutdown");
+#endif
+
+    if (initializeCoreClr == NULL)
+    {
+        printf("coreclr_initialize not found");
+        return -1;
+    }
+
+    if (createManagedDelegate == NULL)
+    {
+        printf("coreclr_create_delegate not found");
+        return -1;
+    }
+
+    if (shutdownCoreClr == NULL)
+    {
+        printf("coreclr_shutdown not found");
+        return -1;
+    }
+
+    //
+    // STEP 3: Construct AppDomain properties used when starting the runtime
+    //
+
+    // Construct the trusted platform assemblies (TPA) list
+    // This is the list of assemblies that .NET Core can load as
+    // trusted system assemblies (similar to the .NET Framework GAC).
+    // For this host (as with most), assemblies next to CoreCLR will 
+    // be included in the TPA list
+    std::string tpaList;
+    BuildTpaList(runtimePath, ".dll", tpaList);
+
+    // <Snippet3A>
+    // Define CoreCLR properties
+    const char* propertyKeys[] = {
+        "TRUSTED_PLATFORM_ASSEMBLIES",      // Trusted assemblies (like the GAC)
+        "APP_PATHS",                        // Directories to probe for application assemblies
```

Comment:
There's no need to specify APP_PATHS, the TPA already resolves all of the assemblies this would help with.
Also we're trying to "deprecate" APP_PATHS and require hosts to always perform full probing if needed.
Same goes for APP_NI_PATHS.
The sample should work just fine without setting this.

Reply:
[CoreRun](https://github.com/dotnet/coreclr/blob/master/src/coreclr/hosts/corerun/corerun.cpp#L563) uses these. Aren't they necessary for hosts that have the managed code they're going to run specified at run time (which I think is not an uncommon scenario)?

Reply:
@vitek-karas I am also confused here.  I thought neither of these were deprecated, but were the  mechanism for passing deps.json info into the runtime

Reply:
CureRun probably does this for historical reasons. Also I would not use CoreRun as a good sample, as it may have things in it to make CoreCLR tests to work which is not necessarily something we want to promote as a "recommended way".

They are not deprecated in any official sense, but in general we want to move to a world where the runtime does not probing. These properties are against that.

Also by default the `dotnet.exe` host will not specify the `APP_PATHS` and `APP_NI_PATHS` properties, in fact you have to know what to look for in the docs to even find the knob which turns it on. I would guess that 99.99% of apps are running with no `APP_PATHS` today. As such our primary hosting sample should not use these either.

`.deps.json` info is passed in `TRUSTED_PLATOFRM_ASSEMBLIES` (the managed code assemblies), `PLATFORM_RESOURCE_ROOTS` (satellite managed assemblies) and `NATIVE_DLL_SEARCH_DIRECTORIES` (the native dependencies), no other properties are used for that. The default host does pass quite a few additional properties, some of which are paths as well, but those are not used by the runtime itself. The properties are currently used as a "hack" to pass information from the host to the managed world (since they are accessible via AppContext class by anyone). Again, that is something we should not "promote" :-)

Reply:
What's the reason for avoiding probing? Is the concern that the wrong assembly could be loaded? Perf related? Perhaps I can add some comments explaining why putting everything on the TPA list is a better approach.

Reply:
I will honestly have to come up with a reasoning around this. The things I know of right now:
* Expectations for people coming from .NET Framework - .NET Framework had a complex probing logic, which we will not port over. Any probing we do have will not work the same and potentially cause confusion.
* Customization - .NET Core has quite a few extensibility mechanisms to customize the assembly resolution process. The probing is part of the runtime and thus inherently not customizable, so any app which would need to customize probing would have to write the whole probing from scratch anyway.
* Ordering - as with the customization above, no matter where in the resolution process we put the probing, there will be a scenario which needs it elsewhere
* Reliability/Trust - some hosts (including our own dotnet.exe) do not perform flat directory scans to find all assemblies, instead they rely on metadata produced by the app build (.deps.json in our case). This means that copying a code file next to the app will not automatically load that code - the app keeps behaving predictably. Probing in the runtime basically defeats that.
* Cross-platform - .NET Core supports using different assemblies on different platforms - technically it's possible to have a single app, which runs on both Windows and Linux and it has OS specific assemblies in it. The definition of which assemblies to use where is done in the .deps.json. Probing will not help here, in fact it would only cause trouble, as it may find the same assembly multiple times, or pick the wrong version.

I completely understand that there are good reasons to have probing, especially for apps/devs coming from .NET Framework, but I think it's better to not have those mechanisms in the runtime itself. The general .NET Core goal is for the runtime to have only the minimum logic around assembly resolution, and provide everything else on top (either from host, or through extensions) - probing is one such piece of logic which doesn't feel like the "minimum".

Reply:
I see what you mean about probing for assemblies not being the 'minimum' functionality required in the runtime. How about I leave the APP_PATH usage but add this comment about the property (and a similar but more fleshed-out one in the associated docs):
```
// Loading from the TPA list is preferred (since the host has more control 
// over what is loaded), but this property can allow probing for user code
// in specified paths.
```

Reply:
Here's a proposed update for how I could described the APP_PATHS property in the docs:

* `APP_PATHS` This is a list of paths to probe in for an assembly if it can't be found in the trusted platform assemblies (TPA) list. Because the host has more control over which assemblies are loaded using the TPA list, it is a best practice for hosts to determine which assemblies they expect to load and list them explicitly. If probing at runtime is needed, however, this property can enable that scenario.

Reply:
The description to put into docs looks good. I personally would still remove it from the sample though - if nothing else, it's not necessary and thus makes the sample a bit confusing (in addition to our desire for people to not use this too much).

That said, if you feel strongly we should have it in the sample... I can live with it :-)

Reply:
That sounds like a good middle ground. I've removed APP_PATHS from the sample, but left it in the docs (with the updated description I mentioned earlier).